mkosi/config.py

```diff
@@ -1809,44 +1810,153 @@ def parse(self, argv: Optional[Sequence[str]] = None) -> tuple[MkosiArgs, tuple[
 
 
 class GenericVersion:
+    _EQUAL = 0
+    _RIGHT_SMALLER = 11
+    _LEFT_SMALLER = 12
+
     def __init__(self, version: str):
         self._version = version
 
+    @classmethod
+    def compare_versions(cls, v1: str, v2: str) -> int:
+        """Implements comparison according to UAPI Group Version Format Specification"""
+        def rstrip_invalid_version_chars(s: str) -> str:
+            valid_version_chars ={*string.ascii_letters, *string.digits, "~", "-", "^", "."}
+            for i, c in enumerate(s):
+                if c in valid_version_chars:
+                    return s[i:]
+            return ""
+
+        def has_either_ended(s1: str, s2: str) -> Optional[int]:
+            if not v1 and not v2:
+                return cls._EQUAL
+            elif not v1 and v2:
+                return cls._LEFT_SMALLER
+            elif v1 and not v2:
+                return cls._RIGHT_SMALLER
+            return None
+
+        def digit_prefix(s: str) -> str:
+            return "".join(takewhile(lambda c: c.isdecimal(), s))
+
+        def letter_prefix(s: str) -> str:
+            return "".join(takewhile(lambda c: c.isalpha(), s))
+
+        while True:
+            #breakpoint()
+            # Any characters which are outside of the set of listed above (a-z, A-Z, 0-9, -, ., ~,
+            # ^) are skipped in both strings. In particular, this means that non-ASCII characters
+            # that are Unicode digits or letters are skipped too.
+            v1 = rstrip_invalid_version_chars(v1)
+            v2 = rstrip_invalid_version_chars(v2)
+            # If the remaining part of one of strings starts with ~: if other remaining part does
```

Comment:
Is "~:" intended here?

Reply:
Yeah, makes more sense, just copied the text from the spec verbatim and the formatting got lost.